src/api/contextMenu/ContextMenuAPI.js

```diff
@@ -75,9 +75,16 @@ class ContextMenuAPI {
     /**
      * @private
      */
-    _showContextMenuForObjectPath(objectPath, x, y, actionsToBeIncluded) {
-
-        let applicableActions = this._allActions.filter((action) => {
+    _showContextMenuForObjectPath(objectPath, x, y, options = {}) {
+        const {
+            actionsToBeIncluded = null,
+            actionsToBeSkipped = []
```

Comment:
I think that we should keep all logic for deciding whether an action should appear in a context menu inside `ContextMenuAction.appliesTo()`. If it's spread around the code it's going to make it really hard for developers to debug. 

Here's a thought - The `PreviewAction` can only be invoked once at a time. Could we simply set a flag on it when it's invoked, and clear it when it's dismissed? Then we could check the flag in `PreviewAction.appliesTo()`? Would save us having to add extra code to the ContextMenuAPI?

Reply:
Was able to make it work, but looks like this solution is little complicated due to following issues:
1. we create new instance of `PreviewAction`, so new flag `this.visible` is not shared (to solve this, created singleton of `PreviewAction`)
1. but now `ViewHistoricalDataAction` is extending `PreviewAction` so `hideInDefaultMenu` flag in causing undesired behavior in `PreviewAction` instance.

Here is PR for that change, have a look at it (may be there is better way?): https://github.com/nasa/openmct/pull/2985

Reply:
> we create new instance of PreviewAction, so new flag this.visible is not shared (to solve this, created singleton of PreviewAction)

You're right, the flag needs to be a singleton. It could be a static property of PreviewAction. Ie on the PreviewAction class itself, not the instance.

> ViewHistoricalDataAction is extending PreviewAction so hideInDefaultMenu flag in causing undesired behavior in PreviewAction instance.

I don't quite understand the interaction here. Does making the `visible` flag a static rather than an instance property help?

Reply:
> I don't quite understand the interaction here. Does making the `visible` flag a static rather than an instance property help?

This was due to singleton class.

Updated as per suggestion, really good suggestion, code looks lot cleaner. 😃